nav2_map_server/src/map_saver.cpp

```diff
@@ -52,105 +53,205 @@ MapSaver::MapSaver(const rclcpp::NodeOptions & options)
 : Node("map_saver", options), save_next_map_promise{}
 {
   Magick::InitializeMagick(nullptr);
-  {
-    mapname_ = declare_parameter("output_file_no_ext", "map");
-    if (mapname_ == "map") {
-      mapname_ += "_" + std::to_string(static_cast<int>(now().seconds()));
-    }
-    threshold_occupied_ = declare_parameter("threshold_occupied", 65);
-    if (100 < threshold_occupied_) {
-      throw std::runtime_error("Threshold_occupied must be 100 or less");
-    }
-    threshold_free_ = declare_parameter("threshold_free", 25);
-    if (threshold_free_ < 0) {
-      throw std::runtime_error("Free threshold must be 0 or greater");
-    }
-    if (threshold_occupied_ <= threshold_free_) {
-      throw std::runtime_error("Threshold_free must be smaller than threshold_occupied");
-    }
 
-    std::string mode_str = declare_parameter("map_mode", "trinary");
+  map_topic_ = declare_parameter("map_topic", "map");
+
+  save_parameters_.map_file_name = declare_parameter("output_file_no_ext", "map");
+  if (save_parameters_.map_file_name == "map") {
+    save_parameters_.map_file_name += "_" + std::to_string(static_cast<int>(now().seconds()));
+  }
+  save_parameters_.occupied_thresh = declare_parameter("threshold_occupied",
+    OCCUPIED_THRESH_DEFAULT);
+
+  save_parameters_.free_thresh = declare_parameter("threshold_free",
+    FREE_THRESH_DEFAULT);
+
+  std::string mode_str = declare_parameter("map_mode", "trinary");
+  try {
+    save_parameters_.mode = map_mode_from_string(mode_str);
+  } catch (std::invalid_argument &) {
+    save_parameters_.mode = MapMode::Trinary;
+    RCLCPP_WARN(
+      get_logger(), "Map mode parameter not recognized: '%s', using default value (trinary)",
+      mode_str.c_str());
+  }
+
+  save_parameters_.image_format = declare_parameter("image_format", save_parameters_.mode == MapMode::Scale ? "png" : "pgm");
+
+  // Checking map parameters for consistency
+  checkSaveParameters(save_parameters_);
+
+  // Make map topic subscription
+  map_sub_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
+    map_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(&MapSaver::mapCallback, this, std::placeholders::_1));
+
+  RCLCPP_INFO(get_logger(), "Waiting for the map");
+}
+
+MapSaver::MapSaver()
```

Comment:
Can you explain the 2 constructors thing? It seems like you just need 1 and what you need are 2 `getParameters()` methods, one that gets from the parameter server and the other that gets from the service request

Reply:
Done. Revised and changed it in 84cff7d.